### PR TITLE
[PX-2202] Parse GPP Header feature and tests.

### PR DESCRIPTION
**Pull request by kylehalverson**
[JIRA Link](https://liveramp.atlassian.net/browse/PX-2202)

- Functionality to Parse GPP Header.
- Tests cases implemented from [IAB's Global Privacy Platform](https://github.com/InteractiveAdvertisingBureau/Global-Privacy-Platform/blob/259d5d4d05c23b2a5d49abf936a2f732be128e1a/Core/Consent%20String%20Specification.md)
  - Please note that the IAB Page has two differing test headers.  My tests use the ones under [GPP String Examples](https://github.com/InteractiveAdvertisingBureau/Global-Privacy-Platform/blob/259d5d4d05c23b2a5d49abf936a2f732be128e1a/Core/Consent%20String%20Specification.md#gpp-string-examples) and NOT [GPP String Format](https://github.com/InteractiveAdvertisingBureau/Global-Privacy-Platform/blob/259d5d4d05c23b2a5d49abf936a2f732be128e1a/Core/Consent%20String%20Specification.md#gpp-string-format).
  - We have a request out to IAB to clarify which set of Headers are expected to use.  In the meantime, I wanted to get 👀 on this.

### CODE REVIEW

File: gpp_parsed_consent.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	var g = &GppHeader{}
+	g.Type, _ = r.ReadInt(6)
+	if g.Type != 3 {
+		return nil, errors.New("wrong gpp header type " + fmt.Sprint(g.Type))
+	}
+	g.Version, _ = r.ReadInt(6)
+	if g.Version != 1 {
+		return nil, errors.New("unsupported gpp version " + fmt.Sprint(g.Version))
+	}
+	g.Sections, _ = r.ReadFibonacciRange()
```
Comment (imang1) [inline gpp_parsed_consent.go:42]: nit that I missed on the last review, but the comment for the func `ReadFibonacciRange` says `ReadFibonacciRangeEntries`

File: gpp_parsed_consent.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	Type     int
+	Version  int
+	Sections []int
+}
+
+// ParseGppHeader parses the first (and required) part of any GPP Consent String.
+// It is used to read the Type, Version, and which sections are contained in the following string(s).
+// Format is:
+// Type	    Int(6)	Fixed to 3 as “GPP Header field”
+// Version	Int(6)	Version of the GPP spec (currently 1)
```
Comment (imang1) [inline gpp_parsed_consent.go]: instead of "currently" can you include the date jan 2023? like Version 1 as of Jan 2023

**Comment by imang1** [comment]
two minor things! Is it possible to hold merging this until we get IAB confirmation on the correct string? How much would that block other work?

File: gpp_parsed_consent.go (hunk 1 above)
Comment (kylehalverson) [reply gpp_parsed_consent.go:42]: Ah, good catch. Updated!

File: gpp_parsed_consent.go (hunk 2 above)
Comment (kylehalverson) [reply gpp_parsed_consent.go]: Updated!

**Comment by kylehalverson** [comment]
> two minor things! Is it possible to hold merging this until we get IAB confirmation on the correct string? How much would that block other work?

It would block the parsing of the entire GPP string. I do need a function that parses the header, checks various things in the entire GPP string, and then parses the different sections.  In the meantime, I can work on the parsing of the US National String, as this is not blocked until we want to call the Parsing function in. that greater ParseGpp string (that combines parsing the header, checking segments vs. sections, and then calling parse functions for each section).

**Review by imang1 — APPROVED** [review]

**Review by imang1 — APPROVED** [review]

File: gpp_parsed_consent_test.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+func (s *GppParseSuite) TestParseGppHeaderFail(c *check.C) {
+	var tcs = []struct {
+		description string
+		header      string
+		expected    error
+	}{
+		{
+			description: "GPP Header must be 3, as of Jan. 2023.",
+			// []byte{0b00000100, 0b00010000, 0b00000010, 0b00110101, 0b10000000}
+			header:   "BBACNYA",
```
Comment (gowthamkommineni) [inline gpp_parsed_consent_test.go:72]: Nit: Do we wanna leave out the last A for test strings. The code adds em anyways.

Comment (kylehalverson) [reply gpp_parsed_consent_test.go:72]: I would likes tests with both to show that they arrive at the same error. Added tests, PTAL

Comment (gowthamkommineni) [reply gpp_parsed_consent_test.go:72]: Cool!

**Review by gowthamkommineni — APPROVED** [review]

**Review by imang1 — APPROVED** [review]

**Pull request merged**